server/src/main/java/org/opensearch/common/util/concurrent/InternalThreadContextWrapper.java

```diff
@@ -0,0 +1,34 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.util.concurrent;
+
+import java.util.Objects;
+
+/**
+ * Wrapper around the ThreadContext to expose methods to the core repo without
+ * exposing them to plugins
+ *
+ * @opensearch.internal
+ */
+public class InternalThreadContextWrapper {
```

Comment:
@cwperks definitely +1 for the intent but I don't think the taking implementation path is correct:
 - The usage of the `ThreadContext` within core should be frictionless. The `InternalThreadContextWrapper` does break this promise, plus the class exposes `public static from` so anyone could do `InternalThreadContextWrapper.from(threadPool.getThreadContext());` to gain the access
 - The usage of the `ThreadContext` outside of the core should not be possible. The long team solution for that is using JPMS but meanwhile, we could close it up using security permissions:
```
private static final Permission ACCESS_SYSTEM_THREAD_CONTEXT_PERMISSION = new RuntimePermission("markAsSystemContext");
```

```
SecurityManager sm = System.getSecurityManager();
        if (sm != null) {
            sm.checkPermission(ACCESS_SYSTEM_THREAD_CONTEXT_PERMISSION);
        }
```

Yes, the compile time checks won't be enforced but runtime checks will be.

Reply:
@reta That sounds like a good approach to grant/prohibit the usage of certain methods in this class in the plugin ecosystem.

I decided to take this approach initially to take advantage of the `@opensearch.internal` annotation so that this repo could use `InternalThreadContextWrapper.from(threadPool.getThreadContext());`, but plugins cannot. I will look into making this a permission that is granted through the JSM policy file.

Reply:
@reta I opened 2 other PRs similar to this one, but they don't make use of this class.

Do you think the other 2 PRs should be updated similarly, or would the changes in those PRs be ok?

- https://github.com/opensearch-project/OpenSearch/pull/14987
- https://github.com/opensearch-project/OpenSearch/pull/14983

Reply:
> Do you think the other 2 PRs should be updated similarly, or would the changes in those PRs be ok?

Thanks @cwperks , I think if we cannot cleanly seal the ThreadContext methods, we would need same change there (plus, changing the method visibility for public API is breaking).

Reply:
I wasn't able to find any usages of those 2 methods outside of the core:

- `stashWithOrigin`: https://github.com/search?q=org%3Aopensearch-project+stashWithOrigin&type=code
- `stashAndMergeHeaders`: https://github.com/search?q=org%3Aopensearch-project%20stashAndMergeHeaders&type=code

Reply:
I mean those are public, right? We sadly don't know who might have been using them (since we don't host all plugins)

Reply:
It would be safe to change the modifier on [`stashAndMergeHeaders`](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/common/util/concurrent/ThreadContext.java#L190-L213) because a plugin would still have access to `stachContext` after that PR and could refactor to 2 lines.

```
final ThreadContext.StoredContext storedContext = threadContext.stashContext();
threadContext.putTransient(ACTION_ORIGIN_TRANSIENT_NAME, origin);
```

I think `stashAndMergeHeaders` would be safe to change based on the usages I see in the core repo, but yes generally changing an access modifier could be breaking.

Reply:
> I think `stashAndMergeHeaders` would be safe to change based on the usages I see in the core repo, but yes generally changing an access modifier could be breaking.

That's the thing, thank you